tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,560 @@
+
+ - Some golden test libraries also automatically updates the golden file.
+
+ - It is useful when working with JSON, HTML, images, etc.
+
+In this tutorial we will use [hspec-golden](http://hackage.haskell.org/package/hspec-golden) library to build some golden-test. We will also use the hspec-golden CLI to easily update our tests.
+
+You can either code along with this tutorial or better undestand it by looking at the finished code. You can find this [code](https://github.com/stackbuilders/tutorials/tree/jguerra-hspec-tutorial) in Github. Feel free to use it.
+
+## Hspec-golden
+
+Hspec-golden is a stackbuilders testing module written in haskell that helps users implement golden test. It is an OpenSource project so feel free to [review](https://github.com/stackbuilders/hspec-golden) it.
+
+## Getting Started
+
+First things first, lets create a new Haskell project on which we will be working:
+
+```bash
+stack new "hspec-golden-tests"
+```
+
+and add the necessary dependencies in order to follow this tutorial without issues.
+
+Project dependencies
+
+``` yaml
+dependencies:
+- base >= 4.7 && < 5
+- aeson
+- text
+- bytestring
+- blaze-html
+```
+
+Testing dependencies
+
+``` yaml
+
+dependencies:
+- hspec-golden-tests
+- hspec
+- hspec-golden
+
+```
+## A Simple Test
+
+Now that we have all dependencies installed, lets create our golden tests.
+
+First lets create a ```HelloWorld ``` module to be tested and add the "Hello Golden Testers" function described earlier.
+
+```haskell
+module HelloWorld where
+
+sayHi :: String
+sayHi = "Hello Golden Testers"
+
+```
+Before moving on with testing,lets review some of hspec-golden [documentation](http://hackage.haskell.org/package/hspec-golden-0.1.0.1/docs/Test-Hspec-Golden.html).
+
+Hspec-golden provides us with a defaultGolden function which creates and compares the SUT output and the golden file. It takes two parameters:
+
+  - the name of the test (Directory where our golden file will be located)
+  - the output of our SUT 
+
+```haskell
+
+defaultGolden :: String -> String -> Golden String
+
+```
+
+By default this function search for golden tests inside a "/.golden" directory, so it would be a great idea if we create one right now.
+
+```shell 
+
+.
+├── .golden
+├── app
+├── src
+└── test
+
+```
+
+Ok enough reading, lets create our tests. We will first create a test module named ```HelloGoldenSpec``` . For convenience I have created this module under test/Hello directory.
+
+```Haskell
+
+module Hello.HelloGoldenSpec where
+
+```
+
+We need some imports.
+
+```Haskell
+
+import           Test.Hspec              -- Test
+import           Test.Hspec.Golden       -- Golden Tests  
+import           HelloWorld              -- SUT
+
+```
+Finally lets write the test.
+
+```Haskell
+
+spec :: Spec
+spec =
+    describe "sayHi" $
+    it "returns Hello Golden Testers string" $
+    defaultGolden "hello" sayHi
```

Comment:
@Jagl257 Let's indent the examples, too.

Reply:
Ok Cris, done.